dask/array/_shuffle.py

```diff
@@ -0,0 +1,199 @@
+from __future__ import annotations
+
+from itertools import count, product
+
+import numpy as np
+
+from dask import config
+from dask.array.chunk import getitem
+from dask.array.core import Array, unknown_chunk_message
+from dask.base import tokenize
+from dask.highlevelgraph import HighLevelGraph
+
+
+def shuffle(x, indexer: list[list[int]], axis):
+    """
+    Reorders one dimensions of a Dask Array based on an indexer.
+
+    The indexer defines a list of positional groups that will end up in the same chunk
+    together. A single group is in at most one chunk on this dimension, but a chunk
+    might contain multiple groups to avoid fragmentation of the array.
+
+    The algorithm tries to balance the chunksizes as much as possible to ideally keep the
+    number of chunks consistent or at least manageable.
+
+    Parameters
+    ----------
+    x: dask array
+        Array to be shuffled.
+    indexer:  list[list[int]]
+        The indexer that determins which elements along the dimension will end up in the
+        same chunk. Multiple groups can be in the same chunk to avoid fragmentation, but
+        each group will end up in exactly one chunk.
```

Comment:
This behaviour is problematic for

```python
import xarray as xr
import dask.array

time = xr.date_range("2001-01-01", "2010-01-01", freq="D")
ds = xr.Dataset(
    {"data": (("x", "y", "time"), dask.array.random.random((10000, 1000, len(time)), chunks=(-1, -1, 1)))},
    coords={"time": ("time", time)},
)
ds.resample(time="Y")._group_indices
```

Reply:
Can you say a bit more?

Is the problem that the chunks will become too large or that we fuse multiple groups together?

If the latter, then we could add control for this behavior to turn fusion of groups off.

If chunks that are too large:
Would your preferred alternative be to have one group in multiple chunks along the time dimension in your example or split the other dimensions into multiple chunks instead of -1 and -1?

Reply:
> one group in multiple chunks along the time dimension in your

I feel like its preferrable to do this and then have the user do the rechunk-to-one-block-while-adjusting-others.

Reply:
Is the user in this case xarray or users from xarray?

When we put one group into multiple chunks, all chunks that have parts of the group should be in consecutive chunks, so that 

x = x.rechunk(("x": "foo", "y": "bar", "time": 365})

would put all values of said group into a single chunk in your example?

Reply:
^ yes. Once the groups are in order, we can use slicing to extract each group and pass to the user.

Basically it'd be nice to preserve the cases where the map approach works nicely today.

```
array.resample(time="Y").reduce(np.mean)
```

Today this dispatches to `dask.array.mean` with each group split across multiple chunks, so we get the tree reduction.

Reply:
Yep ok, got it. This is what you would use take for after https://github.com/dask/dask/pull/11267 (what we discussed in the other issue). take uses the same algorithm under the hood but is not placing everything into a single group.